daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/JsonLedgerClient.scala

```diff
@@ -471,33 +509,56 @@ object JsonLedgerClient {
         s"Request to $path with ${reqBody.map(_.compactPrint)} failed with status $respStatus: $errors"
       )
 
+  // Explicit party specifications for command submissions
+  final case class SubmitParties(
+      actAs: OneAnd[Set, Ref.Party],
+      readAs: Set[Ref.Party],
+  )
+
+  // Expect party specifications for queries
+  final case class QueryParties(
+      readers: OneAnd[Set, Ref.Party]
+  )
+
+  // Validate that the token has the required claims and return
+  // SubmitParties we need to pass to the JSON API
+  // if the token has more claims than we need.
   def validateSubmitParties(
       actAs: OneAnd[Set, Ref.Party],
       readAs: Set[Ref.Party],
       tokenPayload: AuthServiceJWTPayload,
-  ): Either[String, Unit] = {
+  ): Either[String, Option[SubmitParties]] = {
+    val actAsSet = actAs.toList.toSet[String]
+    val readAsSet = readAs.toSet[String]
     val tokenActAs = tokenPayload.actAs.toSet
-    // Relax once the JSON API supports multi-party read/write
-    import scalaz.std.string._
-    val onlyReadAs = readAs.diff(actAs.toSet)
-    val tokenOnlyReadAs = tokenPayload.readAs.toSet.diff(tokenActAs)
+    val tokenReadAs = tokenPayload.readAs.toSet
+    val missingActAs = actAs.toSet.map(p => p: String) diff tokenActAs
+    val missingReadAs = readAs.map(p => p: String) diff (tokenReadAs union tokenActAs)
     if (tokenPayload.actAs.isEmpty) {
       Left(
         s"Tried to submit a command with actAs = [${actAs.toList.mkString(", ")}] but token contains no actAs parties."
       )
 
-    } else if (actAs.toList.toSet[String] /== tokenActAs) {
+    } else if (!missingActAs.isEmpty) {
       Left(
         s"Tried to submit a command with actAs = [${actAs.toList.mkString(", ")}] but token provides claims for actAs = [${tokenPayload.actAs
-          .mkString(" ")}]"
+          .mkString(", ")}]. Missing claims: [${missingActAs.mkString(", ")}]"
       )
-    } else if (onlyReadAs.toSet[String] /== tokenOnlyReadAs) {
+    } else if (!missingReadAs.isEmpty) {
       Left(
         s"Tried to submit a command with readAs = [${readAs.mkString(", ")}] but token provides claims for readAs = [${tokenPayload.readAs
-          .mkString(" ")}]"
+          .mkString(", ")}]. Missing claims: [${missingReadAs.mkString(", ")}]"
       )
     } else {
-      Right(())
+      import scalaz.std.string._
+      val onlyReadAs = readAsSet diff actAsSet
+      val tokenOnlyReadAs = tokenReadAs diff tokenActAs
+      if (onlyReadAs === tokenOnlyReadAs && actAsSet == tokenActAs) {
```

Comment:
Should it not be  
```suggestion
      if (onlyReadAs === tokenOnlyReadAs && actAsSet === tokenActAs) {
```
?